ca_tools/scripts/pos_controller_pid.py

```diff
@@ -0,0 +1,212 @@
+from tf.transformations import euler_from_quaternion
+
+
+class DrawSquare:
+    """
+    Class that contains all the functions of the program.
+    """
+    def __init__(self):
+        """
+        Initialization of the variables that are going to be used in the class.
+        """
+
+        #ROS variables
+
+        self.twist_publisher = rospy.Publisher('/create1/cmd_vel', Twist, queue_size=10)
+        self.odom_subscriber = rospy.Subscriber("/create1/gts", Odometry, self.callback)
+
+        # Time variables
+
+        self._now = 0.0
+        self._time_change = 0.0
+        self._last_time = 0.0
+
+        # Angle variables
+
+        self._current_angle = 0.0
+        self._set_angle = 0.0
+        self._input_angle = 0.0
+        self._error_angle = 0.0
+        self._err_sum_angle = 0.0
+        self._output_angle = 0.0
+        self._control_error_angle = 0.01
+
+        # Lenght variables
+
+        self._set_length = Pose2D()
+        self._input_length = 0.0
+        self._error_length = 0.0
+        self._err_sum_length = 0.0
+        self._output_length = 0.0
+        self._control_error_length = 0.1
+
+        # Auxiliar Variables
+
+        self._curr_aux = Pose2D()
+        self._ini_aux = Pose2D()
+        self._fin_aux = Pose2D()
+        self._is_new_goal = Pose2D()
+
+        # State Variables
+
+        self._set_variable = True
+        self._state = "FORWARD"
+
+        # Twist Variable
+        self._vel = Twist()
+
+        # Odometry Flag
+
+        self._odom_flag = False
+
+        # PID variables
+        self._kp = 5
+        self._ki = 2
+        self._kd = 0
+
+        # Range of output
+
+        self._max_output_angle = 0.4
+        self._min_output_angle = 0.4
+
+        self._max_output_length = 0.4
+        self._min_output_length = 0.4
+
+    def callback(self, data):
+        """
+        Callback function to update the pose according to the robot's movement.
+        """
+
+        q = (data.pose.pose.orientation.x,
+             data.pose.pose.orientation.y,
+             data.pose.pose.orientation.z,
+             data.pose.pose.orientation.w)
+
+        _, _, self._current_angle = euler_from_quaternion(q)
+
+        self._curr_aux.x = data.pose.pose.position.x
+        self._curr_aux.y = data.pose.pose.position.y
+
+        self._odom_flag = True
+
+    def get_time(self):
+        """
+        Function to get the current time.
+        """
+
+        self._now = rospy.get_time()
+        if self._last_time < 0.001:
+            self._last_time = self._now
+        self._time_change = self._now - self._last_time
+
+    def update_time(self):
+        """
+        Function to update the Last Time.
+        """
+
+        if self._now > 0:
+            self._last_time = self._now
+
+    def set_goals(self):
+        """
+        Function that sets the new goals checking the state and if the _set_variable flag
+        is turned True, if it is true, means that a new goal needs to be set.
```

Comment:
Wouldn't it be better that this function returns the boolean value, instead of setting global variables?

Reply:
I'm not sure how to approach this.

Reply:
Nevermind, it's good enough. In general we should try to avoid using too many global variables. Use more parameters and return values from the functions, when possible.